File: java/Weed3/src/noear/weed/Variate.java
package noear.weed;

import noear.weed.ext.DbType;

import java.io.Serializable;
import java.math.BigDecimal;
import java.util.Date;

/**
 * Created by noear on 14-6-12.
 * 数据库访问参数（支付范型）
 */
public class Variate implements Serializable {
    protected String _name;
    protected Object _value;
    public final   int   _hash;

    protected  Variate(){
        _hash=0;
    }

    public Variate(String name, Object value) {
        this._name = name;
        this._value = value;

        if (name != null)
            this._hash = name.hashCode();
        else
            this._hash = 0;
    }

    public boolean isNull(){
        return _value == null;
    }

    public String getName() {
        return _name;
    }

    public void setName(String name) {
        _name = name;
    }


    DbType _type = DbType.Object;
    public  DbType getType() {
        return _type;
    }

    public Object getValue() {
        return _value;
    }

    public void setValue(Object value) {
        _value = value;
    }

    //--------------------
    public <T> T value(T def) {
        if (_value == null)
            return def;
        else {
            return (T) _value;
        }
    }

    public double doubleValue(double def){
        if(_value == null){
            return def;
        }

        if(_value instanceof Double){
            return (double)_value;
        }

        if(_value instanceof Float){
            return (double)((float)_value);
        }

        if(_value instanceof BigDecimal){
            return ((BigDecimal)_value).doubleValue();
        }

        if(_value instanceof Long){
            return (long)_value;
        }

        if(_value instanceof Integer){
            return (int)_value;
        }

        if(_value instanceof Date){
            return ((Date)_value).getTime();
        }

        return def;
    }

    public long longValue(long def){
        if(_value == null){
            return def;
        }

        if(_value instanceof Long){
            return (long)_value;
        }

        if(_value instanceof Integer){
            return (int)_value;
        }

        if(_value instanceof BigDecimal){
            return ((BigDecimal)_value).longValue();
        }

        if(_value instanceof Date){
            return ((Date)_value).getTime();
        }

        return def;
    }

    public int intValue(int def){
        if(_value == null){
            return def;
        }

        if(_value instanceof Integer){
            return (int)_value;
        }

        if(_value instanceof Long){
            return (int) ((long)_value);
        }

        if(_value instanceof BigDecimal){
            return ((BigDecimal)_value).intValue();
        }

        return def;
    }

    public String stringValue(String def){
        if(_value == null){
            return def;
        }

        if(_value instanceof String){
            return (String)_value;
        }else{
            return _value.toString();
        }
    }


//    private Object getDefault(DbType type)
//    {
//        if(type == DbType.String || type == type.DateTime)
//            return null;
//        else if(type == DbType.Int64)
//            return 0L;
//        else if(type == DbType.Double || type == DbType.Single)
//            return 0.0;
//        else if(type == DbType.Boolean)
//            return false;
//        else
//            return 0;
//    }
}

File: .net & mono/Weed3/Variate.cs
using System;
using System.Data;

namespace Noear.Weed {
    /**
 * Created by noear on 14-6-12.
 * 数据库访问参数
 */
    public class Variate {
        protected String _name;
        protected object _value;
        internal readonly  int   _hash;

        internal Variate() {
            _hash = 0;
        }

        public Variate(String name, Object value) { 
            this._name = name;
            this._value = value;

            if (name != null)
                this._hash = name.GetHashCode();
        }
        
        public String getName() {
            return _name;
        }

        public void setName(String name) {
            _name = name;
        }

        
        DbType _type = 0;
        public virtual DbType getType() {
            if (_type == 0) {
                var _typeName = _value.GetType().Name;
                _type = DbTypeMapping.GetType(_typeName);
            }
            return _type;
        }
        
        public virtual Object getValue() {
            return _value;
        }

        public virtual void setValue(Object value) {
            _value = value; 
        }

        //--------------------
        public T value<T>(T def) {
            if (_value == null)
                return def;
            else if (DBNull.Value.Equals(_value))
                return def;
            else
                return doValue<T>();
        }
        
        T doValue<T>() {
            try {
                return (T)_value;
            }
            catch (Exception) {
                if (_name != null)
                    throw new WeedException(_name + "::无法将类型为“" + _value.GetType().ToString() + "”的对象强制转换为类型“" + typeof(T).ToString() + "”");
                else
                    throw new WeedException("无法将类型为“" + _value.GetType().ToString() + "”的对象强制转换为类型“" + typeof(T).ToString() + "”");
            }
        }

        public long longValue(long def) {
            if (_value == null) {
                return def;
            }

            if (_value is long){
                return (long)_value;
            }

            if (_value is int){
                return (int)_value;
            }

            if (_value is Decimal){
                return (int)((Decimal)_value);
            }

            return def;
        }

        public int intValue(int def) {
            if (_value == null) {
                return def;
            }

            if (_value is int) {
                return (int)_value;
            }

            if (_value is long){
                return (int)((long)_value);
            }

            if (_value is Decimal){
                return (int)((Decimal)_value);
            }

            return def;
        }

        public double doubleValue(double def) {
            if (_value == null) {
                return def;
            }

            if (_value is double) {
                return (double)_value;
            }

            if (_value is long) {
                return (long)_value;
            }

            if (_value is int) {
                return (int)_value;
            }

            if (_value is Decimal) {
                return (double)((Decimal)_value);
            }

            return def;
        }
    }
}
